[Qi Zhi] iP

riccqi opened the pull request:
## Duke is getting task support!

It now supports these features and more:
- adding todos, events or deadlines
- ~~deleting tasks~~
- find tasks

Steps to **download** it:
1. download it from [here](https://github.com/riccqi/ip/releases/tag/A-Jar)
2. add tasks through the CLI! 🤩 

Commands to use in the `cli`:
- [x] todo
- [x] event
- [x] deadline
- [x] list
- [x] bye
- [x] delete
- [x] mark
- [x] unmark

> Bye!

``` java
public void printGreeting() {
        String logo = " ____        _        \n"
                + "|  _ \\ _   _| | _____ \n"
                + "| | | | | | | |/ / _ \\\n"
                + "| |_| | |_| |   <  __/\n"
                + "|____/ \\__,_|_|\\_\\___|\n";
        System.out.println("Hello from\n" + logo);
    }
```

Commit:
Add Gradle support

riccqi committed:
feat: add greetings to duke

riccqi committed:
feat: add ability to store tasks

riccqi committed:
feat: add ability to mark and unmark tasks

riccqi committed:
refactor: abstracted more string logic using toString

riccqi committed:
feat: add parsing for different tasks

riccqi committed:
test: add testing using input.txt

riccqi committed:
feat: add error handling for duke exceptions

riccqi committed:
feat: add error handling for duke exceptions

riccqi committed:
feat: add ability to delete

riccqi committed:
refactor: add enum support

riccqi committed:
test: fix tests

riccqi committed:
fix: formatting and imports

riccqi committed:
feat: add storage class for saving and loading

riccqi committed:
refactor: add the parser class to parse saved tasks

riccqi committed:
refactor: change all date representations to be in localdatetime

riccqi committed:
Merge branch 'branch-Level-7'

riccqi committed:
Merge branch 'branch-Level-8'

riccqi committed:
refactor: add tasklist class

riccqi committed:
refactor: add ui class and refactor parser class

riccqi committed:
refactor: add packages

riccqi committed:
Merge branch 'add-gradle-support'

riccqi committed:
test: add JUnit tests to task class

riccqi committed:
chore: add javadocs

riccqi committed:
chore: add javadocs for tasklist

riccqi committed:
fix: switch case formatting

riccqi committed:
feat: add ability to find tasks

riccqi committed:
Merge branch 'branch-A-CodingStandard'

riccqi committed:
Merge branch 'branch-Level-9'

riccqi committed:
fix: merge conflict

riccqi committed:
Chore: Add more javadocs to Parser class

sarrrdin commented on README.md:
```diff
@@ -1,19 +1,24 @@
-3. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
+3. After that, locate the `src/main/java/duke.Duke.java` file, right-click it, and choose `Run duke.Duke.main()` (if the
+   code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something
+   like the below as the output:
```
I really like your edits here makes it much more readable 😸

sarrrdin commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,111 @@
+ * The main class of the Duke program.
+ */
+public class Duke {
+    private TaskList tasklist;
```
Maybe tasklist can consider using camel case?
```
taskList
```

riccqi replied:
good catch!

sarrrdin commented on src/main/java/duke/utils/TaskList.java:
```diff
@@ -0,0 +1,107 @@
+
+    public Task getTask(int taskNum) throws Exception {
+        Task task;
+
```
I think for leaving blank spaces part maybe can be more consistent (based on what i saw on CS2103 requirement haha) so since your code mainly does not have black space before the ```try```, for this you can remove the black space too at line 72.

sarrrdin reviewed (approved):
LGTM 😄 
Code is neat and easy to follow and understand, I like the javadocs and comments that really help in comprehending the code in general.
Just a few cosmetic nits to fix 💪

nmtuan2001 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,111 @@
+                case LIST:
+                    tasklist.printTasks();
+                    break;
+                case MARK: {
```
Maybe you can remove these braces since I think you don't need them for case statements?

riccqi replied:
Yup! this was unnecessary, but for case statements that initialised new variables we should still keep them so the compiler knows the scoping

nmtuan2001 commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,136 @@
+     * @return the parsed Task.
+     * @throws DukeException if the saved string does not match any patterns.
+     */
+    static public Task parseSave(String save) throws DukeException {
```
Maybe you can change the variable name `save` to something a little clearer? The textbook does mention something like "Use Nouns for Things and Verbs for Actions" too. Would something `contentToBeSaved` work (kinda long though)?

riccqi committed:
add checkstyle and fix formatting

fix formatting

Vinodjayakumar124 commented on src/test/java/duke/task/TaskTest.java:
```diff
@@ -0,0 +1,34 @@
+import duke.DukeException;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Perhaps you could consider listing the static imports before the non static imports? Not explicitly stated in the textbook (although the example given under 'The ordering of import statements must be consistent.' demonstrates that) but seems to be a convention.

Vinodjayakumar124 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,107 @@
+                    ui.printBye();
+                    return;
+                case LIST:
+                    taskList.printTasks();
```
Do you think it'll help with consistency with the rest of the switch cases, if we changed the method name from printTasks to listTasks?

Vinodjayakumar124 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+        return "[D] " + super.toString() + " (by: " + by.format(formatter) + ")";
+    }
+
+    public String toSaveString() {
```
Perhaps we could differentiate this method from the previous toString further with a longer method name such as toSaveToFileString to allow developers to quickly understand its purpose

Vinodjayakumar124 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,107 @@
+package duke;
+
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.ToDo;
+import duke.utils.Parser;
+import duke.utils.Storage;
+import duke.utils.TaskList;
+import duke.utils.Ui;
```
I really like the filing structure of your project

Vinodjayakumar124 reviewed (commented):
LGTM. The code was really easy to read and understand; neat and well organised. Just identified some nits to fix.

riccqi committed:
feat: add javafx integration

riccqi committed:
fix: improve exception feedback

riccqi committed:
feat: integrate fxml support

riccqi committed:
remove untracked files

riccqi committed:
remove untracked files

riccqi committed:
feat: add assertion checks to duke

riccqi committed:
refactor: abstract constants to parent class

There are many methods throwing Exceptions that are too general, which
may result in checked exceptions being thrown by duke.

Let's refactor the methods to throw DukeException instead.
Using DukeException is preferrable as it standardises the errors
thrown by duke.

We also abstract out the constant date patterns into a parent class Task,
to ensure consistency and reducing code duplication

riccqi committed:
fix: loading save persists is done property

riccqi committed:
chore: add github actions gradle.yml

riccqi committed:
chore: remove checkstyle for testing

riccqi committed:
chore: remove indirection checks

riccqi committed:
feat: add C-DetectDuplicates extension

riccqi committed:
fix fxml and bye command

riccqi committed:
add ui image

riccqi committed:
add user guide in docs folder

riccqi committed:
fix header comments

riccqi committed:
fix file access error

riccqi committed:
add data.txt to gitignore

riccqi committed:
delete data.txt
